File: apps/api-gateway/internal/middleware/metrics.go
package middleware

import (
	"net/http"
	"strconv"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	httpRequestsTotal = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "http_requests_total",
		Help: "Total number of HTTP requests.",
	}, []string{"method", "status"})

	httpRequestDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "http_request_duration_seconds",
		Help:    "HTTP request duration in seconds.",
		Buckets: prometheus.DefBuckets,
	}, []string{"method"})

	httpRequestsInFlight = promauto.NewGauge(prometheus.GaugeOpts{
		Name: "http_requests_in_flight",
		Help: "Number of HTTP requests currently being processed.",
	})

	ProxyRequestsTotal = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "gateway_proxy_requests_total",
		Help: "Total number of proxied requests by service and status.",
	}, []string{"service", "status"})

	ProxyDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "gateway_proxy_duration_seconds",
		Help:    "Proxy request duration in seconds.",
		Buckets: prometheus.DefBuckets,
	}, []string{"service"})
)

// Metrics is an HTTP middleware that records request metrics.
func Metrics(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		httpRequestsInFlight.Inc()
		defer httpRequestsInFlight.Dec()

		sw := &statusWriter{ResponseWriter: w, status: http.StatusOK}
		next.ServeHTTP(sw, r)

		httpRequestsTotal.WithLabelValues(r.Method, strconv.Itoa(sw.status)).Inc()
		httpRequestDuration.WithLabelValues(r.Method).Observe(time.Since(start).Seconds())
	})
}

File: apps/lark-proxy/src/metrics.ts
import { Counter, Histogram, Gauge, Registry, collectDefaultMetrics } from 'prom-client';
import { Hono } from 'hono';
import type { Context, Next } from 'hono';

const register = new Registry();
collectDefaultMetrics({ register });

const httpRequestsTotal = new Counter({
    name: 'http_requests_total',
    help: 'Total HTTP requests',
    labelNames: ['method', 'path', 'status'] as const,
    registers: [register],
});

const httpRequestDuration = new Histogram({
    name: 'http_request_duration_seconds',
    help: 'HTTP request duration in seconds',
    labelNames: ['method', 'path'] as const,
    registers: [register],
});

const httpRequestsInFlight = new Gauge({
    name: 'http_requests_in_flight',
    help: 'Number of HTTP requests currently being processed',
    registers: [register],
});

export async function metricsMiddleware(c: Context, next: Next): Promise<void | Response> {
    httpRequestsInFlight.inc();
    const start = performance.now();
    try {
        await next();
    } finally {
        const duration = (performance.now() - start) / 1000;
        httpRequestsInFlight.dec();
        httpRequestsTotal.inc({ method: c.req.method, path: c.req.path, status: String(c.res.status) });
        httpRequestDuration.observe({ method: c.req.method, path: c.req.path }, duration);
    }
}

export const metricsApp = new Hono();
metricsApp.get('/metrics', async (c) => {
    return c.text(await register.metrics(), 200, { 'Content-Type': register.contentType });
});
